frontend/src/scenes/surveys/surveyLogic.tsx

```diff
@@ -147,7 +148,42 @@ export interface SurveyDateRange {
     date_to: string | null
 }
 
-const getResponseField = (i: number): string => (i === 0 ? '$survey_response' : `$survey_response_${i}`)
+// New function that supports both index-based and ID-based approaches
+export const getResponseFieldWithId = (questionIndex: number, questionId?: string): string[] => {
+    // Always include the index-based key for backward compatibility
+    const keys = [getSurveyResponseKey(questionIndex)]
+
+    // If we have a question ID, also include the ID-based key
+    if (questionId) {
+        keys.push(`$survey_response_${questionId}`)
+    }
+
+    return keys
+}
+
+// Helper function to generate the HogQL condition for checking survey responses in both formats
+export const getResponseFieldCondition = (questionIndex: number, questionId?: string): string => {
+    const fields = getResponseFieldWithId(questionIndex, questionId)
+
+    // For ClickHouse, we need to use coalesce to check both fields
+    // This will return the first non-null value, prioritizing the ID-based format if available
+    return `coalesce(
+        nullIf(JSONExtractString(properties, '${fields[1] || ''}'), ''),
+        nullIf(JSONExtractString(properties, '${fields[0]}'), '')
+    )`
+}
+
+// Helper function to generate the HogQL condition for checking multiple choice survey responses in both formats
+export const getMultipleChoiceResponseFieldCondition = (questionIndex: number, questionId?: string): string => {
+    const fields = getResponseFieldWithId(questionIndex, questionId)
+
+    // For multiple choice, we need to check if either field has a value and use that one
+    return `if(
+        JSONHas(properties, '${fields[1] || ''}') AND length(JSONExtractArrayRaw(properties, '${fields[1] || ''}')) > 0,
+        arrayJoin(JSONExtractArrayRaw(properties, '${fields[1] || ''}')),
```

Comment:
same as https://github.com/PostHog/posthog/pull/29226/files#r1971003950
Also `const value = fields[1]` so you just reuse the value instead of paying the cost of reading from an array multiple times

Reply:
no longer using an array - now an object for easier readability